apps/design/backend/src/language_and_audio/election_strings.ts

```diff
@@ -0,0 +1,330 @@
+      stringInEnglish: party.fullName,
+    }));
+  },
+  [ElectionStringKey.PARTY_NAME](election) {
+    return election.parties.map((party) => ({
+      stringKey: [ElectionStringKey.PARTY_NAME, party.id],
+      stringInEnglish: party.name,
+    }));
+  },
+  [ElectionStringKey.PRECINCT_NAME](election) {
+    return election.precincts.map((precinct) => ({
+      stringKey: [ElectionStringKey.PRECINCT_NAME, precinct.id],
+      stringInEnglish: precinct.name,
+    }));
+  },
+  [ElectionStringKey.STATE_NAME](election) {
+    return [
+      {
+        stringKey: ElectionStringKey.STATE_NAME,
+        stringInEnglish: election.state,
+      },
+    ];
+  },
+};
+
+function getElectionStringConfig({
+  stringKey,
+}: ElectionString): ElectionStringConfig {
+  return typeof stringKey === 'string'
+    ? electionStringConfigs[stringKey]
+    : electionStringConfigs[stringKey[0]];
+}
+
+function extractElectionStrings(election: Election): ElectionString[] {
+  return Object.values(electionStringExtractorFns)
+    .map((extractorFn) => extractorFn(election))
+    .flat();
+}
+
+export async function extractAndTranslateElectionStrings(
+  translator: GoogleCloudTranslator,
+  election: Election
+): Promise<{
+  /** All election strings */
+  electionStrings: UiStringsPackage;
+  /** The subset of election strings to be added to the vxElectionStrings.json file */
+  vxElectionStrings: UiStringsPackage;
+}> {
+  const untranslatedElectionStrings = extractElectionStrings(election);
+  const electionStringsNotToTranslate = untranslatedElectionStrings.filter(
+    (electionString) => {
+      const config = getElectionStringConfig(electionString);
+      return !config.translatable;
+    }
+  );
+  const electionStringsToCloudTranslate = untranslatedElectionStrings.filter(
+    (electionString) => {
+      const config = getElectionStringConfig(electionString);
+      return config.translatable && !config.customTranslationMethod;
+    }
+  );
+  const electionStringsToCustomTranslate = untranslatedElectionStrings.filter(
+    (electionString) => {
+      const config = getElectionStringConfig(electionString);
+      return config.translatable && config.customTranslationMethod;
+    }
+  );
+
+  const electionStrings: UiStringsPackage = {};
+  const vxElectionStrings: UiStringsPackage = {};
+
+  // Election strings not to translate
+  for (const electionString of electionStringsNotToTranslate) {
+    setUiString(
+      electionStrings,
+      LanguageCode.ENGLISH,
+      electionString.stringKey,
+      electionString.stringInEnglish
+    );
+  }
+
+  // Election strings to cloud translate
+  const stringsInEnglish = electionStringsToCloudTranslate.map(
+    ({ stringInEnglish }) => stringInEnglish
+  );
+  for (const languageCode of Object.values(LanguageCode)) {
+    const stringsInLanguage =
+      languageCode === LanguageCode.ENGLISH
+        ? stringsInEnglish
+        : await translator.translateText(stringsInEnglish, languageCode);
```

Comment:
Is this a thing that we can do in parallel, i.e. using `Promise.all`? Should we? Alternatively, is there a bulk API?

Reply:
The API is actually inherently designed for bulk translation in that you can send an array of text and translate the entire array in one request. We're taking advantage of this already and only making 3 requests, for the 3 target languages. The only place for further parallelization via, say, a `Promise.all` is across the target languages, which I don't think is worth it. Translation is proving to be quite fast, and we also already have a caching mechanism in place to further speed things up

Reply:
I missed the "strings" part, i.e. not "string". Thanks for the clarification!

Reply:
Ah got it, no prob!